src/modules/land_detector/MulticopterLandDetector.cpp

```diff
@@ -131,115 +143,147 @@ void MulticopterLandDetector::_update_params()
+			max_climb_rate = _param_lndmc_z_vel_max.get() * 2.5f;
+		}
 
-	} else {
-		// Adjust max_climb_rate if land_speed is lower than 2x max_climb_rate
-		float max_climb_rate = ((land_speed_threshold * 0.5f) < _param_lndmc_z_vel_max.get()) ? (0.5f * land_speed_threshold) :
-				       _param_lndmc_z_vel_max.get();
-		vertical_movement = fabsf(_vehicle_local_position.vz) > max_climb_rate;
+		vertical_movement = (fabsf(_vehicle_local_position.vz) > max_climb_rate);
 	}
 
+
 	// Check if we are moving horizontally.
-	_horizontal_movement = sqrtf(_vehicle_local_position.vx * _vehicle_local_position.vx
-				     + _vehicle_local_position.vy * _vehicle_local_position.vy) > _param_lndmc_xy_vel_max.get();
+	if (lpos_available && _vehicle_local_position.v_xy_valid) {
+		const Vector2f v_xy{_vehicle_local_position.vx, _vehicle_local_position.vy};
+		_horizontal_movement = v_xy.longerThan(_param_lndmc_xy_vel_max.get());
+
+	} else {
+		_horizontal_movement = false; // not known
+	}
+
+
+	// low thrust: 30% of throttle range between min and hover
+	const float sys_low_throttle = _params.minThrottle + (_params.hoverThrottle - _params.minThrottle) * 0.3f;
+
+	bool ground_contact = (_actuator_controls_throttle <= sys_low_throttle) && !vertical_movement;
 
 	// if we have a valid velocity setpoint and the vehicle is demanded to go down but no vertical movement present,
 	// we then can assume that the vehicle hit ground
-	_in_descend = _is_climb_rate_enabled()
-		      && (_vehicle_local_position_setpoint.vz >= land_speed_threshold);
-	bool hit_ground = _in_descend && !vertical_movement;
+	if (_flag_control_climb_rate_enabled) {
+		vehicle_local_position_setpoint_s vehicle_local_position_setpoint;
+
+		if (_vehicle_local_position_setpoint_sub.update(&vehicle_local_position_setpoint)) {
+			// setpoints can briefly be NAN to signal resets, TODO: fix in multicopter position controller
+			const bool descend_vel_sp = PX4_ISFINITE(vehicle_local_position_setpoint.vz)
+						    && (vehicle_local_position_setpoint.vz >= land_speed_threshold);
+
+			const bool descend_acc_sp = PX4_ISFINITE(vehicle_local_position_setpoint.acceleration[2])
+						    && (vehicle_local_position_setpoint.acceleration[2] >= 100.f);
+
+			_in_descend = descend_vel_sp || descend_acc_sp;
+		}
+
+		// ground contact requires commanded descent until landed
+		if (!_maybe_landed_hysteresis.get_state() && !_landed_hysteresis.get_state()) {
+			ground_contact &= _in_descend;
+		}
+
+	} else {
+		_in_descend = false;
+	}
+
+
+	// When not armed, consider to have ground-contact
+	if (!_armed) {
+		return true;
+	}
+
 
 	// TODO: we need an accelerometer based check for vertical movement for flying without GPS
-	return (_has_low_thrust() || hit_ground) && (!_horizontal_movement || !_has_position_lock())
-	       && (!vertical_movement || !_has_altitude_lock());
+	return ground_contact && !_horizontal_movement && !vertical_movement;
```

Comment:
Summary of the big changes:
- The switch to the acceleration setpoint is handled correctly
- No vertical movement via velocity estimate is a requirement (even in case it's not available, see comment above) while before low thrust was enough